hive-metastore/src/main/java/org/apache/iceberg/hive/HiveMetastoreConnector.java

```diff
@@ -0,0 +1,173 @@
+
+  Map<String, String> hmsEnvContext(String metadataLocation);
+
+  default void setSchema(TableMetadata metadata, Map<String, String> parameters) {
+    parameters.remove(TableProperties.CURRENT_SCHEMA);
+    if (exposeInHmsProperties() && metadata.schema() != null) {
+      String schema = SchemaParser.toJson(metadata.schema());
+      setField(parameters, TableProperties.CURRENT_SCHEMA, schema);
+    }
+  }
+
+  default void setField(Map<String, String> parameters, String key, String value) {
+    if (value.length() <= getMaxHiveTablePropertySize()) {
+      parameters.put(key, value);
+    } else {
+      LOG.warn(
+          "Not exposing {} in HMS since it exceeds {} characters",
+          key,
+          getMaxHiveTablePropertySize());
+    }
+  }
+
+  static void validateTableIsIceberg(Table table, String fullName) {
+    String tableType = table.getParameters().get(BaseMetastoreTableOperations.TABLE_TYPE_PROP);
+    NoSuchIcebergTableException.check(
+        tableType != null
+            && tableType.equalsIgnoreCase(BaseMetastoreTableOperations.ICEBERG_TABLE_TYPE_VALUE),
+        "Not an iceberg table: %s (type=%s)",
+        fullName,
+        tableType);
+  }
+
+  default void persistTable(Table hmsTable, boolean updateHiveTable, String metadataLocation)
+      throws TException, InterruptedException {
+    if (updateHiveTable) {
+      getMetaClients()
+          .run(
+              client -> {
+                MetastoreUtil.alterTable(
+                    client, database(), table(), hmsTable, hmsEnvContext(metadataLocation));
+                return null;
+              });
+    } else {
+      getMetaClients()
+          .run(
+              client -> {
+                client.createTable(hmsTable);
+                return null;
+              });
+    }
+  }
+
+  default StorageDescriptor storageDescriptor(TableMetadata metadata, boolean hiveEngineEnabled) {
+
+    final StorageDescriptor storageDescriptor = new StorageDescriptor();
+    storageDescriptor.setCols(HiveSchemaUtil.convert(metadata.schema()));
+    storageDescriptor.setLocation(metadata.location());
+    SerDeInfo serDeInfo = new SerDeInfo();
+    serDeInfo.setParameters(Maps.newHashMap());
+    if (hiveEngineEnabled) {
+      storageDescriptor.setInputFormat("org.apache.iceberg.mr.hive.HiveIcebergInputFormat");
+      storageDescriptor.setOutputFormat("org.apache.iceberg.mr.hive.HiveIcebergOutputFormat");
+      serDeInfo.setSerializationLib("org.apache.iceberg.mr.hive.HiveIcebergSerDe");
+    } else {
+      storageDescriptor.setOutputFormat("org.apache.hadoop.mapred.FileOutputFormat");
+      storageDescriptor.setInputFormat("org.apache.hadoop.mapred.FileInputFormat");
+      serDeInfo.setSerializationLib("org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe");
+    }
+    storageDescriptor.setSerdeInfo(serDeInfo);
+    return storageDescriptor;
+  }
+
+  default void cleanupMetadata(FileIO fileIO, String commitStatus, String metadataLocation) {
```

Comment:
Could this be static?

Reply:
nit: maybe `io` instead of `fileIO`? - not important, just seems strange to me... 😄